Add Jenkins CI/CD script example (#370)
astronomer/docs: cloud/ci-cd.md

### GitHub Actions

> To automate code deploys to a single Deployment using [GitHub Actions](https://github.com/features/actions), complete the following setup in a GitHub repository hosting an Astronomer project:

Reviewer: ```suggestion
To automate code deploys to a Deployment on Astronomer Cloud using [GitHub Actions](https://github.com/features/actions), complete the following setup in a GitHub repository that hosts an Astronomer project:
```

            --header 'Content-Type: application/json' \
            --data-raw '{"query":"mutation imageDeploy(\n    $input: ImageDeployInput!\n  ) {\n    imageDeploy(\n      input: $input\n    ) {\n      id\n      deploymentId\n      digest\n      env\n      labels\n      name\n      tag\n      repository\n    }\n}","variables":{"input":{"id":"${{ steps.image.outputs.image_id }}","tag":"deploy-${{ steps.date.outputs.date }}","repository":"images.astronomer.cloud/${{ env.ORGANIZATION_ID }}/${{ env.DEPLOYMENT_ID }}"}}}'
    ```

### Jenkins Script

> You can use the provided [Jenkins](https://www.jenkins.io/) script in any Git-based repository that hosts a single Astronomer project created via `astrocloud dev init`.  To do so:

Reviewer: ```suggestion
To automate deploys to Astronomer Cloud with [Jenkins](https://www.jenkins.io/), use the script below in any Git-based repository that hosts an [Astronomer project](create-project.md).
```

> 1. At the root of your repository, add a file names `Jenkinsfile` that includes the following script:

Reviewer: ```suggestion
1. At the root of your repository, add a file named `Jenkinsfile` that includes the following script:
```

    ```jenkins
    pipeline {
     agent any
       stages {

>          stage('Deploy to astronomer') {

Reviewer: ```suggestion
         stage('Deploy to Astronomer') {
```

           when {
            expression {
              return env.GIT_BRANCH == "origin/main"
            }
           }
           steps {
             script {
               sh "chmod +x -R ${env.WORKSPACE}"
               sh('./build.sh')
             }
           }
         }
       }
     post {
       always {
         cleanWs()
       }
     }
    }

>     ```

Reviewer: It'd be good to add a sentence to describe what's going on in this first file

Reviewer 2: @paolaperaza @aspain I have tested the following Jenkinsfile to test passing environment variables and credentials for multiple deployments:
```
pipeline {
 agent any
   environment {
     TEST_API_CREDS = credentials("astro_test_api_key")
     PROD_API_CREDS = credentials("astro_prod_api_key")
   }
   stages {
     stage('Deploy to Test mk-demo') {
       when {
        expression {
          return env.BRANCH_NAME =~ '^(test).*'
        }
       }
       steps {
         script {
           /* sh 'printenv' */
           sh 'echo ${TEST_DEPLOYMENT_ID}'
           sh "chmod +x -R ${env.WORKSPACE}"
           sh('./build.sh ${TEST_DEPLOYMENT_ID} ${TEST_API_CREDS_USR} ${TEST_API_CREDS_PSW}')
         }
       }
     }
     stage('Deploy to Prod mk-demo') {
       when {
        expression {
          return env.BRANCH_NAME =~ '^(prod).*'
        }
       }
       steps {
         script {
           /* sh 'printenv' */
           sh 'echo ${PROD_DEPLOYMENT_ID}'
           sh "chmod +x -R ${env.WORKSPACE}"
           sh('./build.sh ${PROD_DEPLOYMENT_ID} ${PROD_API_CREDS_USR} ${PROD_API_CREDS_PSW}')
         }
       }
     }
   }
 post {
   always {
     cleanWs()
   }
 }
}
```

Reviewer: Thanks @manmeetkaur ! This is great to have for reference, though I think there's not an obvious place to add this example _right_ now.

@jwitz To keep things moving, can you create a new docs GitHub issue to cover, "Add Best Practices for pushing to multiple Deployments with CI/CD" and copy paste this script there? I figure this might end up being another section in this doc entirely

2. At the root of your repository, create a file called `build.sh` and add the following to it:

    ```sh
    # Create time stamp
    TAG=deploy-`date "+%Y-%m-%d-%HT%M-%S"`

>     # Step 1. Use the Access token in the `astro auth login` command

Reviewer: ```suggestion
    # Step 1. Authenticate to Astronomer's Docker registry with your Deployment API key ID and secret. This is equivalent to running `$ astrocloud auth login` via the Astronomer Cloud CLI.
```

    docker login images.astronomer.cloud -u $ASTRONOMER_KEY_ID -p $ASTRONOMER_KEY_SECRET

    # Step 2. Request the Organization Id

> 

Reviewer: Step here?

Author: Ah yeah, just noticed this isn't in the original GitHub issue either. @aspain assuming this can be removed since we define the Org ID as an environment variable later?

>     # Step 3. Build the image

Reviewer: ```suggestion
    # Step 3. Build your Astronomer project into a tagged Docker image.
```

    docker build . -t images.astronomer.cloud/$ORGANIZATION_ID/$DEPLOYMENT_ID:$TAG

>     # Step 4. Push the image

Reviewer: ```suggestion
    # Step 4. Push that Docker image to Astronomer's Docker registry.
```

    docker push images.astronomer.cloud/$ORGANIZATION_ID/$DEPLOYMENT_ID:$TAG

>     # Step 5. Get the access token

Reviewer: ```suggestion
    # Step 5. Fetch an API access token with your Deployment API key ID and secret.
```

    echo "get token"
    TOKEN=$( curl --location --request POST "https://auth.astronomer.io/oauth/token" \
            --header "content-type: application/json" \
            --data-raw "{
                \"client_id\": \"$ASTRONOMER_KEY_ID\",
                \"client_secret\": \"$ASTRONOMER_KEY_SECRET\",
                \"audience\": \"astronomer-ee\",
                \"grant_type\": \"client_credentials\"}" | jq -r '.access_token' )

>     # Step 6. Create the Image

Reviewer: ```suggestion
    # Step 6. Make a request to the Astronomer API that passes metadata from your new Docker image and creates a record for it.

```

    echo "get image id"
    IMAGE=$( curl --location --request POST "https://api.astronomer.io/hub/v1" \
            --header "Authorization: Bearer $TOKEN" \
            --header "Content-Type: application/json" \
            --data-raw "{
                \"query\" : \"mutation imageCreate(\n    \$input: ImageCreateInput!\n) {\n    imageCreate (\n    input: \$input\n) {\n    id\n    tag\n    repository\n    digest\n    env\n    labels\n    deploymentId\n  }\n}\",
                \"variables\" : {
                    \"input\" : {
                        \"deploymentId\" : \"$DEPLOYMENT_ID\",
                        \"tag\" : \"$TAG\"
                        }
                    }
                }" | jq -r '.data.imageCreate.id')

>     # Step 7. Deploy the Image

Reviewer: ```suggestion
    # Step 7. Pass the repository URL for the Docker image to your Astronomer Deployment. This completes the deploy process and triggers your Scheduler and Workers to restart.
```

    echo "deploy image"
    curl --location --request POST "https://api.astronomer.io/hub/v1" \
            --header "Authorization: Bearer $TOKEN" \
            --header "Content-Type: application/json" \
            --data-raw "{
                \"query\" : \"mutation imageDeploy(\n    \$input: ImageDeployInput!\n  ) {\n    imageDeploy(\n      input: \$input\n    ) {\n      id\n      deploymentId\n      digest\n      env\n      labels\n      name\n      tag\n      repository\n    }\n}\",
                \"variables\" : {
                    \"input\" : {
                        \"id\" : \"$IMAGE\",
                        \"tag\" : \"$TAG\",
                        \"repository\" : \"images.astronomer.cloud/$ORGANIZATION_ID/$DEPLOYMENT_ID\"
                        }
                    }
                }"
    ```

3. Add the following environment variables in your repository:

    - `ASTRONOMER_KEY_ID`: Your Deployment API key ID
    - `ASTRONOMER_KEY_SECRET`: Your Deployment API key secret

>     - `ORGANIZATION_ID`: Your organization

Reviewer 3: ```suggestion
    - `ORGANIZATION_ID`: Your organization ID
```

>     - `DEPLOYMENT_ID`: Your Deployment ID

Reviewer: It might not technically matter, but does this step make sense as Step 1 since these Env Vars are referenced in the script above?

Reviewer 3: yeah couldn't hurt to move this earlier in the process

>     - `ORGANIZATION_ID`: Your organization
>     - `DEPLOYMENT_ID`: Your Deployment ID

Reviewer: ```suggestion
    - `ASTRONOMER_ORGANIZATION_ID`: Your organization
    - `ASTRONOMER_DEPLOYMENT_ID`: Your Deployment ID
```

@aspain How would you handle these if you have Jenkins CI for multiple Deployments?

Reviewer 3: I believe you'd just have to add additional variables for each deployment and then edit the Jenkinsfile trigger rules to dictate when to deploy to dev vs prod etc. May need someone more familiar with Jenkins to confirm this 

    - `ASTRONOMER_ORGANIZATION_ID`: Your organization
    - `ASTRONOMER_DEV_DEPLOYMENT_ID`: Your Deployment ID
    - `ASTRONOMER_PROD_DEPLOYMENT_ID`: Your Deployment ID

            --header 'Content-Type: application/json' \
            --data-raw '{"query":"mutation imageDeploy(\n    $input: ImageDeployInput!\n  ) {\n    imageDeploy(\n      input: $input\n    ) {\n      id\n      deploymentId\n      digest\n      env\n      labels\n      name\n      tag\n      repository\n    }\n}","variables":{"input":{"id":"${{ steps.image.outputs.image_id }}","tag":"deploy-${{ steps.date.outputs.date }}","repository":"images.astronomer.cloud/${{ env.ORGANIZATION_ID }}/${{ env.DEPLOYMENT_ID }}"}}}'
    ```

### Jenkins

> To automate code deploys to a single Deployment using [Jenkins](https://www.jenkins.io/), complete the following setup in a Git-based repository hosting an Astronomer project:

Reviewer: ```suggestion
To automate code deploys to a Deployment on Astronomer Cloud using [Jenkins](https://www.jenkins.io/), complete the following setup in a Git-based repository that hosts an Astronomer project:
```

> 1. At the root of your repository, add a file named `Jenkinsfile` that includes the following script:

Author: ```suggestion
1. At the root of your Git repository, add a file named `Jenkinsfile` that includes the following script:
```
```suggestion
1. At the root of your repository, add a file named `Jenkinsfile` that includes the following script:
```

> 1. At the root of your Git repository, add a file named `Jenkinsfile` that includes the following script:

Reviewer 2: A Jenkinsfile is a text file that contains the definition of a Jenkins Pipeline and is checked into source control. It helps writing and maintaining complex pipelines easier which might be difficult to handle in the Script Text Area of a classic Pipeline or a Freestyle Project.

Author: Since this sentence is in the original Jenkins documentation, it's probably better for us to link out instead of copying the sentence ourselves. I updated the step so that the word "Jenkinsfile" links out to the appropriate Jenkins doc.

> 1. At the root of your Git repository, add a [Jenkinsfile](https://www.jenkins.io/doc/book/pipeline/jenkinsfile/) that includes the following script:

Reviewer: ```suggestion
1. Add a [Jenkinsfile](https://www.jenkins.io/doc/book/pipeline/jenkinsfile/) that includes the following script:
```

    ```jenkins
    pipeline {
     agent any
       stages {
         stage('Deploy to Astronomer') {
           when {
            expression {
              return env.GIT_BRANCH == "origin/main"
            }
           }
           steps {
             script {
               sh "chmod +x -R ${env.WORKSPACE}"
               sh('./build.sh')
             }
           }
         }
       }
     post {
       always {
         cleanWs()
       }
     }
    }
    ```

    Specifically, this Jenkinsfile deploys to Astronomer every time there is a merge to the `main` branch of your repository.

> 

Reviewer 2: also, @paolaperaza @aspain passing the args from Jenkinsfile, we need to make the following change to build.sh:
```
DEPLOYMENT_ID=$1
ASTRONOMER_KEY_ID=$2
ASTRONOMER_KEY_SECRET=$3
```

>     Specifically, this Jenkinsfile deploys to Astronomer every time there is a merge to the `main` branch of your repository.

Reviewer: ```suggestion
    This Jenkinsfile triggers a code push to Astronomer Cloud every time a commit or Pull Request is merged to the `origin/main` branch of your repository.
```

> 2. Add the following environment variables in your repository:

Reviewer: ```suggestion
2. Add the following environment variables to your repository:
```

    - `ASTRONOMER_KEY_ID`: Your Deployment API key ID
    - `ASTRONOMER_KEY_SECRET`: Your Deployment API key secret
    - `ORGANIZATION_ID`: Your Organization ID
    - `DEPLOYMENT_ID`: Your Deployment ID

> 3. At the root of your repository, create a file called `build.sh` and add the following to it:

Reviewer 3: @paolaperaza I think it works here, it can really be in any directory as long as the full path is referenced in the `sh('./build.sh')` portion of the Jenkinsfile

>     ```sh
>     # Create time stamp
>     TAG=deploy-`date "+%Y-%m-%d-%HT%M-%S"`

Reviewer: ```suggestion
    ```sh
    DEPLOYMENT_ID=$1
    ASTRONOMER_KEY_ID=$2
    ASTRONOMER_KEY_SECRET=$3
    
    # Create time stamp
    TAG=deploy-`date "+%Y-%m-%d-%HT%M-%S"`
```

Per @manmeetkaur, sounds like we need to add these Env Vars to the top of the `build.sh` file? Do we not need the Organization ID?

Reviewer 2: I have added the organization id in the file itself, or we can add it to the environment variable becasue it is not going to change .
